### PR TITLE
Add support for zod lazy and recursive zod schema

### PR DESCRIPTION
**Pull request by jadedevin13**

### CODE REVIEW

**Comment by brettz9** [comment]
This is exciting to see... Could you:

1. Run `prettier --write src` on the code? Besides enforcing the pre-existing style pattern, it should make the diffs easier to read.
2. Add some tests? (ideally so that `npm test` shows perfect coverage (besides `ui.ts`))

File: src/zerialize.ts [hunk 1]
```diff
@@ -306,62 +315,94 @@ const zerializers = {
...
+  ZodPromise: (def, ctx) => ({ type: 'promise', value: s(def.type, ctx) }),
 
-  ZodLazy: (def) => zerialize(def.getter()),
-  ZodEffects: (def) => zerialize(def.schema),
-  ZodBranded: (def) => zerialize(def.type),
-  ZodPipeline: (def) => zerialize(def.out),
-  ZodCatch: (def) => zerialize(def.innerType),
+  ZodLazy: (def, ctx) => {
+    if (ctx.seenSchemas.has(def.getter())) {
+      return { type: 'lazy', ref: ctx.seenSchemas.get(def.getter()) };
```
Comment (brettz9) [inline src/zerialize.ts:366]: I think in order to remain JSON-compatible in the serialized format, this would need to be a JSON Pointer or such.

Comment (jadedevin13) [reply src/zerialize.ts]: Updated to use refs.

```json
{
  "type": "object",
  "properties": {
    "limit": { "type": "number", "coerce": true, "id": 3, "isOptional": true },
    "offset": { "type": "number", "coerce": true, "id": 5, "isOptional": true },
    "orderBy": {
      "type": "array",
      "element": {
        "type": "object",
        "properties": {
          "id": {
            "type": "object",
            "properties": {
              "isAsc": {
                "type": "boolean",
                "coerce": true,
                "id": 12,
                "isOptional": true
              },
              "isDesc": {
                "type": "boolean",
                "coerce": true,
                "id": 14,
                "isOptional": true
              }
            },
            "id": 10,
            "isOptional": true
          },
          "test": { "type": "ref", "ref": 10 }
        },
        "id": 9
      },
      "id": 7,
      "isOptional": true
    },
    "id": {
      "type": "object",
      "properties": {
        "isNull": {
          "type": "boolean",
          "coerce": true,
          "id": 18,
          "isOptional": true
        },
        "isNotNull": {
          "type": "boolean",
          "coerce": true,
          "id": 20,
          "isOptional": true
        },
        "eq": {
          "type": "string",
          "coerce": true,
          "id": 22,
          "isOptional": true
        },
        "ne": {
          "type": "string",
          "coerce": true,
          "id": 24,
          "isOptional": true
        },
        "gt": {
          "type": "string",
          "coerce": true,
          "id": 26,
          "isOptional": true
        },
        "gte": {
          "type": "string",
          "coerce": true,
          "id": 28,
          "isOptional": true
        },
        "lt": {
          "type": "string",
          "coerce": true,
          "id": 30,
          "isOptional": true
        },
        "lte": {
          "type": "string",
          "coerce": true,
          "id": 32,
          "isOptional": true
        },
        "like": {
          "type": "string",
          "coerce": true,
          "id": 34,
          "isOptional": true
        },
        "notLike": {
          "type": "string",
          "coerce": true,
          "id": 36,
          "isOptional": true
        },
        "ilike": {
          "type": "string",
          "coerce": true,
          "id": 38,
          "isOptional": true
        },
        "notIlike": {
          "type": "string",
          "coerce": true,
          "id": 40,
          "isOptional": true
        },
        "between": {
          "type": "object",
          "properties": {
            "lower": { "type": "string", "coerce": true, "id": 44 },
            "upper": { "type": "ref", "ref": 44 }
          },
          "id": 42,
          "isOptional": true
        },
        "notBetween": {
          "type": "object",
          "properties": {
            "lower": { "type": "string", "coerce": true, "id": 47 },
            "upper": { "type": "ref", "ref": 47 }
          },
          "id": 45,
          "isOptional": true
        }
      },
      "id": 16,
      "isOptional": true
    },
    "test": { "type": "ref", "ref": 16 },
    "and": {
      "type": "array",
      "element": {
        "type": "lazy",
        "schema": {
          "type": "lazy",
          "schema": {
            "type": "object",
            "properties": {
              "id": { "type": "ref", "ref": 16 },
              "test": { "type": "ref", "ref": 16 },
              "and": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "schema": { "type": "ref", "ref": 51 },
                  "id": 55
                },
                "id": 53,
                "isOptional": true
              },
              "or": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "ref": { "type": "ref", "ref": 51 },
                  "id": 58
                },
                "id": 56,
                "isOptional": true
              }
            },
            "id": 52
          },
          "id": 51
        },
        "id": 50
      },
      "id": 48,
      "isOptional": true
    },
    "or": {
      "type": "array",
      "element": {
        "type": "lazy",
        "ref": {
          "type": "lazy",
          "schema": {
            "type": "object",
            "properties": {
              "id": { "type": "ref", "ref": 16 },
              "test": { "type": "ref", "ref": 16 },
              "and": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "schema": { "type": "ref", "ref": 51 },
                  "id": 55
                },
                "id": 53,
                "isOptional": true
              },
              "or": {
                "type": "array",
                "element": {
                  "type": "lazy",
                  "ref": { "type": "ref", "ref": 51 },
                  "id": 58
                },
                "id": 56,
                "isOptional": true
              }
            },
            "id": 52
          },
          "id": 51
        },
        "id": 61
      },
      "id": 59,
      "isOptional": true
    }
  },
  "id": 1
}

```

Comment (brettz9) [reply src/zerialize.ts]: Thanks for the response. However, I think part of the appeal of allowing a JSON serialization format is that it should be easy to craft from JSON as well, so, imo, I really think JSON references would be a much better solution.

Comment (jadedevin13) [reply src/zerialize.ts]: In case you change your mind let me know. I updated the test and docs.

**Comment by Gregoor** [comment]
Great, thank you! It’s too long to review on my phone, so I’ll try to remember to look at it next week when I’m at my laptop again

File: src/zerialize.ts (hunk 1 above)
Comment (brettz9) [reply src/zerialize.ts]: Not my call. Will see what @Gregoor decides...

**Comment by brettz9** [comment]
@Gregoor : I've added #29 as an alternative to this PR. It avoids requiring `id` on each object, instead doing its own internal tracking of paths. It also uses JSON references as the pointers which, besides being familiar to JSON Schema and other users, can optionally be processed first with another JSON references-aware library to allow modular storage of schemas in other files.

**Comment by brettz9** [comment]
I think this can be closed now that #29 landed.

**Pull request closed without merging**